evm-contracts/src/v0.7/dev/Operator.sol

```diff
@@ -175,6 +234,55 @@ contract Operator is
     return success;
   }
 
+  /**
+   * @notice Called by the Chainlink node to fulfill requests with multi-word support
+   * @dev Given params must hash back to the commitment stored from `oracleRequest`.
+   * Will call the callback address' callback function without bubbling up error
+   * checking in a `require` so that the node can get paid.
+   * @param requestId The fulfillment request ID that must match the requester's
+   * @param payment The payment amount that will be released for the oracle (specified in wei)
+   * @param callbackAddress The callback address to call for fulfillment
+   * @param callbackFunctionId The callback function ID to use for fulfillment
+   * @param expiration The expiration that the node should respond by before the requester can cancel
+   * @param data The data to return to the consuming contract
+   * @return Status if the external call was successful
+   */
+  function fulfillOracleRequest2(
+    bytes32 requestId,
+    uint256 payment,
+    address callbackAddress,
+    bytes4 callbackFunctionId,
+    uint256 expiration,
+    bytes memory data
+  )
+    external
+    override
+    onlyAuthorizedNode()
+    isValidRequest(requestId)
+    returns (bool)
+  {
+    _verifyOracleResponse(
+      requestId,
+      payment,
+      callbackAddress,
+      callbackFunctionId,
+      expiration
+    );
+    emit OracleResponse2(
+      requestId,
+      payment,
+      callbackAddress,
+      callbackFunctionId,
+      expiration,
+      data);
+    require(gasleft() >= MINIMUM_CONSUMER_GAS_LIMIT, "Must provide consumer enough gas");
+    // All updates to the oracle's fulfillment should come before calling the
+    // callback(addr+functionId) as it is untrusted.
+    // See: https://solidity.readthedocs.io/en/develop/security-considerations.html#use-the-checks-effects-interactions-pattern
+    (bool success, ) = callbackAddress.call(abi.encodeWithSelector(callbackFunctionId, requestId, data)); // solhint-disable-line avoid-low-level-calls
```

Comment:
Meant to include this in the last review. This will not work for a single word response, because `data` always takes up at least two words(length and payload).

In order to efficiently use `encodeWithSelector`, we need `data` to always include `requestId`. After that, it can encode any number of parameters of varying length, but if we assume that they have been properly encoded before being passed in then we don't have to make any modifications. Some examples of what it could look like:
- `bytes memory data = abi.encode(requestId,someInt256)` (single word A)
- `bytes memory data = abi.encode(requestId,someBytes32)` (single word B)
- `bytes memory data = abi.encode(requestId,someInt256,someBytes32)` (multiword A)
- `bytes memory data = abi.encode(requestId,someUint256Array)` (multiword B, `uint256[]` is equivalent to an abi encoded payload being `bytes`)
- `bytes memory data = abi.encode(requestId,someUserDefinedStruct)` (multiword C)
...

I see three options for checking request ID: We can pass `requestId` as an explicit parameter into `fulfillOracleRequest2` and check that it matches the first 32 bytes of data. Alternatively, if we pass a `requestId` parameter in, we can not check data but instead just overwrite the first 32 bytes. Finally, we could not pass it in as its own parameter, but write a method to extract it from `data`.

Ideally we'd decide on this by comparing optimized gas usage for each one. For now, I'd say we can start with any of them along with good tests. Then we can refactor and optimize later.

Reply:
Thanks for adding more context Steve. 

It looks like I've made some false assumptions about what multi-word enables consumer contracts. I assumed that it simply enables consumers to implement a fulfilment function in the same format as currently, but with bytes as a return value: `(bytes32 _requestId, bytes memory _price)`.

I'll get digging into the first option: pass `requestId` as an explicit parameter and checking that it matches the first 32 bytes of data.